test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,501 @@
+end)
+
+-- Demoting current follower should succeed.
+g_nspaces.test_follower_demote = function(g)
+    local ok, err = g.server_2:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while demoting follower with elections off: %s', err))
+
+    g.server_1:box_config({election_mode = 'manual'})
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, string.format(
+        'error while demoting follower with elections on: %s', err))
+end
+
+-- Promoting current raft leader should succeed,
+-- even if he doesn't own synchro queue with elections enabled.
+g_nspaces.test_raft_leader_promote = function(g)
+    g.server_1:box_config({election_mode = 'manual'})
+
+    -- Promote server, but get stuck before obtaining synchro queue
+    -- (write term bump to wal, get stuck on promote)
+    wal_delay_start(g.server_1, 2)
+    g.server_1:exec(function()
+        box.ctl.promote()
+    end)
+    g.server_1:wait_wal_delay()
+    g.server_1:wait_election_state('leader')
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+
+    luatest.assert(ok, string.format(
+        'error while promoting raft leader: %s', err))
+end
+
+g_nspaces.after_test('test_raft_leader_promote', function(g)
+    -- Finish promoting server_1
+    wal_delay_end(g.server_1)
+    g.server_1:wait_synchro_queue_owner_id()
+
+    g.server_1:box_config({election_mode = 'off'})
+    g.cluster:demote(g.server_1)
+    g.server_2:wait_synchro_queue_owner_id(0)
+end)
+
+-- Promoting and demoting should work when everything is ok.
+g_nspaces.test_ok = function(g)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    g.server_2:wait_synchro_queue_owner_id(g.server_1:instance_id())
+    luatest.assert(ok, string.format(
+        'error while promoting with elections off: %s', err))
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    g.server_2:wait_synchro_queue_owner_id(0)
+    luatest.assert(ok, string.format(
+        'error while demoting with elections off: %s', err))
+
+    g.server_1:box_config({election_mode = 'manual'})
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    g.server_2:wait_synchro_queue_owner_id(g.server_1:instance_id())
+    luatest.assert(ok, string.format(
+        'error while promoting with elections on: %s', err))
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    g.server_2:wait_lsn(g.server_1)
+    luatest.assert(ok, string.format(
+        'error while demoting with elections on: %s', err))
+end
+
+g_nspaces.after_test('test_ok', function(g)
+    g.server_1:box_config({election_mode = 'off'})
+    g.cluster:promote(g.server_2)
+    g.cluster:demote(g.server_2)
```

Comment:
Are you sure this promote + demote pair is needed? It looks like a workaround for waiting for some state, or something.

You demote `server1` right in the test, so this promote + demote shouldn't be necessary.

Reply:
Thanks for review! Yeah, demoted it, with election_mode='manual', between tests election_mode is set to 'off', and we will have #6860. I have no real idea how to fix it here without promote/demote pair. I can add a comment on why that's the case.

Reply:
Thanks for the explanation! I see now.
Ok, let it be then, but please add some comment mentioning this issue